core/services/periodicbackup/backup.go

```diff
@@ -0,0 +1,139 @@
+package periodicbackup
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/url"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/smartcontractkit/chainlink/core/logger"
+)
+
+var (
+	fileName           = "db_backup.tar.gz"
+	minBackupFrequency = time.Minute
+)
+
+type backupResult struct {
+	size int64
+	path string
+}
+
+type (
+	DatabaseBackup interface {
+		Start() error
+		Close() error
+		RunBackupGracefully()
+	}
+
+	databaseBackup struct {
+		logger          *logger.Logger
+		databaseURL     url.URL
+		frequency       time.Duration
+		outputParentDir string
+		done            chan bool
+	}
+)
+
+func NewDatabaseBackup(frequency time.Duration, databaseURL url.URL, outputParentDir string, logger *logger.Logger) DatabaseBackup {
+	return &databaseBackup{
+		logger,
+		databaseURL,
+		frequency,
+		outputParentDir,
+		make(chan bool),
+	}
+}
+
+func (backup databaseBackup) Start() error {
+
+	ticker := time.NewTicker(backup.frequency)
+
+	go func() {
+		for {
+			select {
+			case <-backup.done:
+				ticker.Stop()
+				return
+			case <-ticker.C:
+				if backup.frequencyIsTooSmall() {
+					logger.Errorf("Database backup frequency (%s=%v) is too small. Please set it to at least %s", "DATABASE_BACKUP_FREQUENCY", backup.frequency, minBackupFrequency)
+					continue
+				}
+				backup.RunBackupGracefully()
+			}
+		}
+	}()
+
+	return nil
+}
+
+func (backup databaseBackup) Close() error {
+	backup.done <- true
+	return nil
+}
+
+func (backup *databaseBackup) frequencyIsTooSmall() bool {
+	return backup.frequency < minBackupFrequency
+}
+
+func (backup *databaseBackup) RunBackupGracefully() {
+	backup.logger.Info("DatabaseBackup: Starting database backup...")
+	startAt := time.Now()
+	result, err := backup.runBackup()
+	duration := time.Since(startAt)
+	if err != nil {
+		backup.logger.Errorf("DatabaseBackup: Failed after %s with: %v", duration, err)
+	} else {
+		backup.logger.Infof("DatabaseBackup: Database backup finished successfully after %s: %d bytes written to %s", duration, result.size, result.path)
+	}
+}
+
+func (backup *databaseBackup) runBackup() (*backupResult, error) {
+
+	tmpFile, err := ioutil.TempFile(backup.outputParentDir, "db_backup")
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to create a tmp file")
+	}
+	err = os.Remove(tmpFile.Name())
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to remove the tmp file before running backup")
+	}
+
+	cmd := exec.Command(
+		"pg_dump", backup.databaseURL.String(),
+		"-f", tmpFile.Name(),
+		"-F", "t", // format: tar
```

Comment:
This is going to be incredibly expensive. We do not want to run regular full pg_dumps on all tables, it will slow the node down badly.

Ideally this would dump from a read replica. Can we pass in an additional optional ENV var like `BACKUP_DATABASE_URL` or something?

Regardless, we should be very careful about what we dump. Can we explicitly include only certain, small tables here instead?

Reply:
ok created optional BACKUP_DATABASE_URL now